src/parser/core/modules/ActionWindow/evaluators/RulePassedEvaluator.tsx

```diff
@@ -0,0 +1,44 @@
+import {RotationTarget} from 'components/ui/RotationTable'
+import React, {ReactNode} from 'react'
+import {Icon} from 'semantic-ui-react'
+import {Suggestion} from '../../Suggestions'
+import {EvaluatedAction} from '../EvaluatedAction'
+import {HistoryEntry} from '../History'
+import {NotesEvaluator} from './NotesEvaluator'
+
+export interface RulePassedEvaluatorOpts {
```

Comment:
Style point: Not personally a huge fan of how this essentially is "derive from RulePassedEvaluator by passing in a bunch of function objects or not" vs the more standard "derive from RulePassedEvaluator that has virtual methods with default implementations".

(I realize other evaluators take some function objects, but I think this struck me differently than those due to these functions feeling more core logic of the class as compared to handle edge cases in the other evaluators.)

I see why you did it that way, since you need data from the Leylines window object.  The alternative to have RulesPassedEvaluator use regular virtual methods instead would mean that the derived StayedInEvaluator would have to take an `owner: Leylines` and call the methods that way.  Personally, I'm used to doing things that way from other projects I've done, but I'm not sure what the general preference is for this project.

Reply:
I wrote RulesPassed as an abstract Evaluator at first but that ended up meaning a lot additional boilerplate for each extension (5 in BLM, 3 in DNC, so far, plus others I haven't PRed yet) and yes, frequently having to pass quite a lot of data in through constructor properties to let it do the checks, which is why I eventually gravitated to this model.

This comment made me take a closer look at other places we'd eventually want to use it to replace/polish up all the places we're showing a column with the checkmark for success and X for failure, and I think you're right that having it be an abstract other things can inherit from is probably the better long-term model. I'll swap back to that.